pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1120 @@
+	/// * Else, resolves opposite swap directions by immediately fulfilling the
+	///   side with lower amounts; or both if the swap amounts are equal.
+	///
+	/// Throws if the decreasing amount exceeds the amount which is
+	/// currently swapping into pool currency and/or investing as we cannot
+	/// decrease more than was invested. We must ensure, this can never happen
+	/// at this stage!
+	///
+	/// NOTE: We can ignore handling all states which include
+	/// `SwapIntoForeignDone` without `ActiveSwapIntoForeignCurrency` as we
+	/// consume the done amount and transition in the post transition phase.
+	/// Moreover, we can ignore handling all states which do not include
+	/// `ActiveSwapIntoPoolCurrency` or `InvestmentOngoing` as we cannot reduce
+	/// further then.
+	/// To be safe and to not make any unhandled assumptions, we throw
+	/// `DispatchError::Other` for these states though we need to make sure
+	/// this can never occur!
+	fn handle_decrease(
+		&self,
+		swap: Swap<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		if swap.currency_in == swap.currency_out {
+			return Self::handle_decrease_non_foreign(self, swap);
+		}
+
+		match &self {
+			// Cannot reduce if there is neither an ongoing investment nor an active swap into pool
+			// currency
+			Self::NoState
+			| Self::ActiveSwapIntoForeignCurrency { .. }
+			| Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone { .. } => Err(DispatchError::Other(
+				"Invalid invest state when transitioning a decrease",
+			)),
+			// Increment foreign swap amount up to ongoing investment
+			Self::InvestmentOngoing { invest_amount } => {
+				let foreign_amount_pool_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_out,
+					swap.currency_in,
+					swap.amount,
+				)?;
+
+				match foreign_amount_pool_denominated.cmp(invest_amount) {
+					Ordering::Less => Ok(Self::ActiveSwapIntoForeignCurrencyAndInvestmentOngoing {
+						swap,
+						invest_amount: invest_amount.ensure_sub(foreign_amount_pool_denominated)?,
+					}),
+					Ordering::Equal => Ok(Self::ActiveSwapIntoForeignCurrency { swap }),
+					// should never occur but let's be safe here
+					Ordering::Greater => Err(DispatchError::Arithmetic(ArithmeticError::Underflow)),
+				}
+			}
+			// Increment return done amount up to amount of the active pool swap
+			Self::ActiveSwapIntoPoolCurrency { swap: pool_swap } => {
+				swap.ensure_currencies_match(pool_swap, false)?;
+
+				let foreign_amount_pool_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_out,
+					swap.currency_in,
+					swap.amount,
+				)?;
+
+				match foreign_amount_pool_denominated.cmp(&pool_swap.amount) {
+					Ordering::Less => Ok(Self::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDone {
+						swap: Swap {
+							amount: pool_swap
+								.amount
+								.ensure_sub(foreign_amount_pool_denominated)?,
+							..*pool_swap
+						},
+						done_amount: swap.amount,
+					}),
```

Comment:
Weird case I want to double-check @William,

`swap.amount` used in line 608 is foreign currency denominated, but the resulting `ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDone` I think should contain the `done_amount` denominated as pool currency because the swap is a pool swap.

I would say line 608 should be:
```rust
done_amount: foreign_amount_pool_denominated
```
instead.